cranelift/codegen/src/isa/x64/lower.rs

```diff
@@ -2839,6 +2945,138 @@ fn lower_insn_to_regs<C: LowerCtx<I = Inst>>(
             }
         }
 
+        Opcode::Splat => {
+            let ty = ty.unwrap();
+            assert_eq!(ty.bits(), 128);
+            let src_ty = ctx.input_ty(insn, 0);
+            assert!(src_ty.bits() < 128);
+            let src = input_to_reg_mem(ctx, inputs[0]);
+            let dst = get_output_reg(ctx, outputs[0]);
+
+            fn emit_insert_lane<C: LowerCtx<I = Inst>>(
+                ctx: &mut C,
+                src: RegMem,
+                dst: Writable<Reg>,
+                lane: u8,
+                ty: Type,
+            ) {
+                if !ty.is_float() {
+                    let (sse_op, w_bit) = match ty.lane_bits() {
+                        8 => (SseOpcode::Pinsrb, false),
+                        16 => (SseOpcode::Pinsrw, false),
+                        32 => (SseOpcode::Pinsrd, false),
+                        64 => (SseOpcode::Pinsrd, true),
+                        _ => panic!("Unable to insertlane for lane size: {}", ty.lane_bits()),
+                    };
+                    ctx.emit(Inst::xmm_rm_r_imm(sse_op, src, dst, lane, w_bit));
+                } else if ty == types::F32 {
+                    let sse_op = SseOpcode::Insertps;
+                    // Insert 32-bits from replacement (at index 00, bits 7:8) to vector (lane
+                    // shifted into bits 5:6).
+                    let lane = 0b00_00_00_00 | lane << 4;
+                    ctx.emit(Inst::xmm_rm_r_imm(sse_op, src, dst, lane, false));
+                } else if ty == types::F64 {
+                    let sse_op = match lane {
+                        // Move the lowest quadword in replacement to vector without changing
+                        // the upper bits.
+                        0 => SseOpcode::Movsd,
+                        // Move the low 64 bits of replacement vector to the high 64 bits of the
+                        // vector.
+                        1 => SseOpcode::Movlhps,
+                        _ => unreachable!(),
+                    };
+                    // Here we use the `xmm_rm_r` encoding because it correctly tells the register
+                    // allocator how we are using `dst`: we are using `dst` as a `mod` whereas other
+                    // encoding formats like `xmm_unary_rm_r` treat it as a `def`.
```

Comment:
I'm a little uneasy about the hidden semantics here -- it feels like a leaky abstraction. (I know these are pre-existing, but the confusion is appearing now that I see this comment...). Could we perhaps have an enum indicating whether the output of a r/m, r instruction depends on the initial value or not (i.e., `IsModOrDef::Def` vs. `IsModOrDef::Mod`)? That would make things more clear, I think.

Reply:
I agree that this is a leaky abstraction but I'm not confident `IsModOrDef` plugs the hole; let me open an issue to discuss: https://github.com/bytecodealliance/wasmtime/issues/2252.